ticker.EngFormatter: allow offset

doronbehar opened the pull request:
## PR summary

Solve #28463 by making `ticker.EngFormatter` be a subclass of `ticker.ScalarFormatter` (and not upon `ticker.Formatter`).

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

tacaswell commented on doc/users/next_whats_new/EngFormatter-offset.rst:
```diff
@@ -0,0 +1,9 @@
+data, and shows the offset with the known SI quantity prefixes. To disable this
+new behavior, simply pass ``useOffset=False`` when you instantiate it. If offsets
+are disabled, or if there is no particular offset that fits your axis data, the
+formatter will reside to the old behavior.
```
```suggestion
formatter will revert to the old behavior.
```

doronbehar replied:
Fixed :).

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1404,114 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
             To enable/disable the use mathtext for rendering the numbers in
             the formatter.
+        useOffset : bool or float, default: :rc:`axes.formatter.useoffset`
+            Whether to use offset notation. See `.set_useOffset`.
```
This is an API change in that Engineering Formatters did not used to even support offsets, but now they default to using them.  I think the behavior here should be: 

 - accept {True, False, None}
 - default to False
 - None means "look at rcparams"

tacaswell commented on doc/users/next_whats_new/EngFormatter-offset.rst:
```diff
@@ -0,0 +1,9 @@
+data, and shows the offset with the known SI quantity prefixes. To disable this
+new behavior, simply pass ``useOffset=False`` when you instantiate it. If offsets
+are disabled, or if there is no particular offset that fits your axis data, the
+formatter will revert to the old behavior.
```
This is also a place where you can throw in a code example via the `.. plot` directive to show off the new feature.

doronbehar replied:
> This is also a place where you can throw in a code example via the `.. plot` directive to show off the new feature.

I added an example instead.

doronbehar replied:
> you can also copy/paste the example code/simplify it into the what's new - it doesn't need the be distinct, the aim is just to advertise the new functionality

I think it will be better advertised in the examples, but if you think it is not worth an example I won't mind do as you suggested.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -407,6 +407,9 @@ class ScalarFormatter(Formatter):
         See `.set_useLocale`.
+    usetex : bool, default: :rc:`text.usetex`
+        To enable/disable the use of TeX's math mode for rendering the
+        numbers in the formatter.
```
This (and useOffset below) should get a `.. versionadded 3.10`

doronbehar commented:
Thanks for all the suggestions! I hope the documentation related comments will come out OK.

tacaswell commented:
Does that change the behavior when the user does not opt-in to using offsets?

doronbehar commented:
OK so after thinking about it once more, I realized that what I implemented in the demo at https://github.com/matplotlib/matplotlib/issues/28463 may be too complex - there is no need to show an order of magnitude of the data at the offset text, if we already have the `ENG_PREFIXES` that can shorten the strings of the ticks. Hence I managed to further simplify the logic in `EngFormatter` such that `self.orderOfMagnitude` is always 0, and the offset text only shows offset prefixed with `+`, and not showing the oom of the data after offset subtraction.

The only debatable drawback of this behavior, is that a `0` tick will always appear without a unit prefix. It sort of makes sense because 0 is 0 no matter what is the data's general oom. I got convinced after playing a bit, that making sure that all ticks have the same unit prefix is making the class so much more complex that it doesn't worth it. The current behavior is consistent and well tested.

The last force push simplifies as described above, and improves the test I added for this feature.

doronbehar commented:
OK so after testing many more edge cases, I revised the behavior of this I changed my opinion once more. The tests now are much more extensive and I finally managed to reach the behavior I was aiming for. 

The git history got too messy and I had a few reverts locally so I forced push.

Here are a few screenshots:

### +/-0.5e6

![image](https://github.com/matplotlib/matplotlib/assets/10998835/6ef31696-c5f4-4a85-9b1b-24342233e8b0)

### +/- 3.1e3

![image](https://github.com/matplotlib/matplotlib/assets/10998835/2e4431a5-b4ec-4087-94d1-6b0c0243d88e)

### +/- 2.1e5

![image](https://github.com/matplotlib/matplotlib/assets/10998835/3b4fc8eb-f725-45f3-a9df-707924c165e1)

### Also it is automatically updated when zooming in

![image](https://github.com/matplotlib/matplotlib/assets/10998835/1768837a-0d3c-4975-8db7-21465c5b7d8e)

### Offset enabled in principal, but data doesn't qualify to get an offset:

(Also when started like that, offset is computed automatically and presented if needed)

![image](https://github.com/matplotlib/matplotlib/assets/10998835/360ab51f-82d4-4b5b-8f16-292b7933916c)

doronbehar commented:
I almost lost in the rebase the fixes to the comments mentioned above. They are fixed now. Hopefully docs will build fine now :pray:.

doronbehar commented:
I'm having some trouble building the documentation locally, so I have to rely upon CI. Last force push fixed a few small issues with it.

doronbehar commented on doc/users/next_whats_new/EngFormatter-offset.rst:
```diff
@@ -0,0 +1,12 @@
+To enable this new behavior, simply pass ``useOffset=True`` when you instantiate
+:class:`matplotlib.ticker.EngFormatter`. Example is available here_.
+
+.. _here: ../../gallery/ticks/engformatter_offset.html
```
> if docs don't build, you probably want something like:
> 
> ```
> :doc:`gallery\ticks\engformatter_offset`
> ```
> 
> [matplotlib.org/devdocs/devel/document.html](https://matplotlib.org/devdocs/devel/document.html#refer-to-other-documents-and-sections)

Thanks! I'm not used to `rst`..

doronbehar replied:
> no problem, I think I gave you the wrong slash direction - updated now.

Thanks again.

> Also you don't need the link...but using the plot directive would mean no link. See how the other what's news for examples:
> [`main`/doc/users/next_whats_new](https://github.com/matplotlib/matplotlib/tree/main/doc/users/next_whats_new?rgh-link-date=2024-07-04T15%3A34%3A28Z)

Yes I'm familiar with the `.. plot` directive, I think still that the examples are better place to document this, because the source code is also available without having to download it.

doronbehar replied:
Aish that link doesn't work according to CI...

https://output.circle-artifacts.com/output/job/8eb8464f-376d-45ab-ad59-ed86f4b29834/artifacts/0/doc/build/html/users/next_whats_new/EngFormatter-offset.html

doronbehar replied:
> > , I think still that the examples are better place to document this,
> 
> To be clear, it's not an either or, it's a thank you very much for the example please also add a plot to the release notes.

I added a `..plot` directive pointing to the same example, as the [devel/document.html](https://matplotlib.org/devdocs/devel/document.html#include-figures-and-files) demonstrate.

doronbehar commented on doc/users/next_whats_new/EngFormatter-offset.rst:
```diff
@@ -0,0 +1,12 @@
+To enable this new behavior, simply pass ``useOffset=True`` when you instantiate
+:class:`matplotlib.ticker.EngFormatter`. Example is available here_.
+
+.. _here: :doc:`gallery/ticks/engformatter_offset`
```
> And underneath you can also link to the example via plot_directive to embed the image [matplotlib.org/devdocs/devel/document.html](https://matplotlib.org/devdocs/devel/document.html#include-figures-and-files)

Thanks, the link helped, and I used the examples there so probably it will work now, let's see.

doronbehar commented:
Docs look great and all links work :tada: :

https://output.circle-artifacts.com/output/job/4914ad26-73c1-467e-9d8f-66ba6621e491/artifacts/0/doc/build/html/users/next_whats_new/engformatter_offset.html

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -444,20 +449,31 @@ class ScalarFormatter(Formatter):
     """
 
-    def __init__(self, useOffset=None, useMathText=None, useLocale=None):
+    def __init__(self, useOffset=None, useMathText=None, useLocale=None, usetex=None):
```
```suggestion
    def __init__(self, useOffset=None, useMathText=None, useLocale=None, *, usetex=None):
```

If we are adding this, might as well add it key-word only to start.

doronbehar replied:
> If we are adding this, might as well add it key-word only to start.

TBH I'm not sure I understand this subtlety, but I did this change anyway.

doronbehar commented:
The CI test failure on Uubuntu with Python3.11 seems unrelated to this PR...

tacaswell commented:
```python
def foo(a, b=None):
    ...
```

can be called as

```python
foo(1)
foo(1, 2)
foo(1, b=2)
foo(a=1, b=2)
```

```python
def bar(a, *, b=None):
    ...
```

can be called as

```python
bar(1)
bar(1, b=2)
```

for these toy examples, this isn't a bug deal, but a call signature like `obj = MyClass('a', True, True, False, None)` is much less readable.  Forcing the new parameters to be keyword only helps future readability and lets us re-order the parameters if we want to (see https://docs.python.org/3/tutorial/controlflow.html#special-parameters).

ksunden commented:
Cycling to rerun tests (since we no longer have logs)

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -407,6 +407,11 @@ class ScalarFormatter(Formatter):
     useLocale : bool, default: :rc:`axes.formatter.use_locale`.
         Whether to use locale settings for decimal sign and positive sign.
         See `.set_useLocale`.
+    usetex : bool, default: :rc:`text.usetex`
```
It seems like this should get a What's New entry.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
+            guaranteed to get an offset which will make the tick labels exceed
+            3 digits. See also `.set_useOffset`.
+
+        .. versionadded:: 3.10
```
Should be same level as parameter description
```suggestion
            .. versionadded:: 3.10
```

QuLogic commented on lib/matplotlib/ticker.pyi:
```diff
@@ -64,8 +64,14 @@ class ScalarFormatter(Formatter):
     ) -> None: ...
     offset: float
+    def get_usetex(self) -> bool: ...
+    def set_usetex(self, val: bool | float) -> None: ...
```
Not `float` AFAICT
```suggestion
    def set_usetex(self, val: bool) -> None: ...
```

QuLogic commented on lib/matplotlib/ticker.pyi:
```diff
@@ -64,8 +64,14 @@ class ScalarFormatter(Formatter):
     offset: float
+    def get_usetex(self) -> bool: ...
+    def set_usetex(self, val: bool | float) -> None: ...
+    @property
+    def usetex(self) -> bool: ...
```
This needs a setter type hint, as it's not read-only.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
-        if val is None:
-            self._useMathText = mpl.rcParams['axes.formatter.use_mathtext']
+    def __call__(self, x, pos=None):
+        """
+        Return the format for tick value *x* at position *pos*. If there is no
+        currently offset in the data, it returns the best engineering formatting
+        that fits the given argument, independently.
+        """
```
This sentence doesn't read right; hopefully this is correct:
```suggestion
        """
        Return the format for tick value *x* at position *pos*.
        
        If there is currently no offset in the data, it returns the best engineering
        formatting that fits the given argument, independently.
        """
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1468,13 +1532,15 @@ def format_eng(self, num):
             pow10 += 3
 
-        prefix = self.ENG_PREFIXES[int(pow10)]
+        unitPrefix = self.ENG_PREFIXES[int(pow10)]
+        if self.unit or unitPrefix:
+            suffix = f"{self.sep}{unitPrefix}{self.unit}"
```
```suggestion
        unit_prefix = self.ENG_PREFIXES[int(pow10)]
        if self.unit or unit_prefix:
            suffix = f"{self.sep}{unit_prefix}{self.unit}"
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1468,13 +1532,15 @@ def format_eng(self, num):
+            return rf"${mant:{fmt}}${suffix}"
         else:
-            formatted = f"{mant:{fmt}}{self.sep}{prefix}"
-
-        return formatted
+            return rf"{mant:{fmt}}{suffix}"
```
Why are these raw strings now? There are no backslashes?

doronbehar replied:
> Why are these raw strings now? There are no backslashes?

Agree.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
+    def set_locs(self, locs):
+        # docstring inherited
+        self.locs = locs
+        if len(self.locs) > 0:
```
```suggestion
        if self.locs:
```

doronbehar replied:
The `len(self.locs) > 0` -> `self.locs` was a mistake, you get a `ValueError: The truth value of an array with more than one element is ambiguous.` error.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
-        return self.fix_minus(s)
+    # Simplify a bit ScalarFormatter.get_offset: We always want to use
+    # self.format_data. Also we want to return a non-empty string only if there
+    # is an offset, no matter what is self.orderOfMagnitude. if there is an
```
```suggestion
    # is an offset, no matter what is self.orderOfMagnitude. If there is an
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
+    # Simplify a bit ScalarFormatter.get_offset: We always want to use
+    # self.format_data. Also we want to return a non-empty string only if there
+    # is an offset, no matter what is self.orderOfMagnitude. if there is an
+    # offset OTH, self.orderOfMagnitude is consulted. This behavior is verified
```
OTH?

doronbehar replied:
> OTH?

It sounded good to me in the sentence to use that word, but it can be neglected :).

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
+            return ''
+        if self.offset:
+            offsetStr = ''
+            if self.offset:
```
This is already true based on the `if` two lines above; is there something missing in this condition?

doronbehar replied:
> This is already true based on the `if` two lines above; is there something missing in this condition?

I'm pretty sure not, the logic behaves as I want according to the tests I defined, so I'll just simplify that part to behave the same way.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
+                    sciNotStr = r'\times%s' % sciNotStr
+                s = fr'${sciNotStr}{offsetStr}$'
+            else:
+                s = ''.join((sciNotStr, offsetStr))
```
Seems complicated:
```suggestion
                s = sciNotStr + offsetStr
```

doronbehar replied:
Right, thanks.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1390,76 +1407,123 @@ def __init__(self, unit="", places=None, sep=" ", *, usetex=None,
+            if self._useMathText or self._usetex:
+                if sciNotStr != '':
+                    sciNotStr = r'\times%s' % sciNotStr
+                s = fr'${sciNotStr}{offsetStr}$'
```
Why a raw string?

doronbehar replied:
> Why a raw string?

Not an aware choice - it is a remnant from old code.

QuLogic reviewed (commented):
Mostly minor stuff, but a couple of small implementation questions.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1594,6 +1594,79 @@ def test_engformatter_usetex_useMathText():
 
 
+@pytest.mark.parametrize(
+    'oom_center, oom_noise, oom_center_desired, oom_noise_desired', [
```
I would write `oom_center` and `oom_noise` here as their actual values, instead of `log10` of them. Then it should (hopefully) be clear where in the relative order of magnitude should be. You can write them grouped in 3s such as:
```
        (271_490_000_000, 10, 9, 0),
        (27_149_000_000_000, 10_000_000, 12, 6),
        ...
```

doronbehar replied:
> I would write `oom_center` and `oom_noise` here as their actual values, instead of `log10` of them (and modify the test to not do `10**oom_{center,noise}`.)
> 
> Then it should (hopefully) be clear where in the relative order of magnitude should be. You can write them grouped in 3s such as:
> 
> ```
>         (271_490_000_000, 10, 9, 0),
>         (27_149_000_000_000, 10_000_000, 12, 6),
>         ...
> ```

Hmm but then isn't it a bit weird how `oom_` means something different when it is the first 2 values v.s the second? It is much more intuitive to me how all of these variables are in the same logarithmic scale, and it would be easier to maintain these tests IMO like this (at least for me).

QuLogic replied:
Well yes, you would remove the `oom_` prefix on those ones. The problem with using order-of-magnitude to me is that it seems a bit circular, just from looking at the parametrization.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1594,6 +1594,79 @@ def test_engformatter_usetex_useMathText():
+):
+    UNIT = "eV"
+    # Doesn't really matter here, but should be of order of magnitude ~= 1
+    r = range(-5, 7)
```
```suggestion
    r = np.arange(-5, 7)
```

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1594,6 +1594,79 @@ def test_engformatter_usetex_useMathText():
+    fig, ax = plt.subplots()
+    # Use some random ugly number
+    data_offset = 2.7149*10**oom_center
+    ydata = data_offset + np.array(r, dtype=float)*10**oom_noise
```
```suggestion
    ydata = data_offset + r*10**oom_noise
```

doronbehar replied:
Done something a bit different eventually, I'm pretty sure I got your point.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1594,6 +1594,79 @@ def test_engformatter_usetex_useMathText():
+    formatter.ENG_PREFIXES[0] = "_"
+    ax.yaxis.set_major_formatter(formatter)
+    fig.canvas.draw()
+    offsetGot = formatter.get_offset()
+    ticksGot = [labl.get_text() for labl in ax.get_yticklabels()]
```
Though `ticker.py` is not great at it, please stick to PEP8 variable naming here.
```suggestion
    offset_got = formatter.get_offset()
    ticks_got = [labl.get_text() for labl in ax.get_yticklabels()]
```
(though I would just drop the `_got` entirely as there is no other version.)

doronbehar replied:
> (though I would just drop the `_got` entirely as there is no other version.)

That's another thing that helped me debug things with `pytest --showlocals`...

doronbehar replied:
> > (though I would just drop the `_got` entirely as there is no other version.)
> 
> That's another thing that helped me debug things with `pytest --showlocals`...

So I am marking this review thread as resolved, as I got the impression that the `--showlocals` dilemma is resolved.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1594,6 +1594,79 @@ def test_engformatter_usetex_useMathText():
+    # only makes sure that offset text and the ticks gets the correct unit
+    # prefixes and the ticks.
+    if formatter.offset:
+        prefix_noise_got = offsetGot[2]
+        prefix_noise_desired = formatter.ENG_PREFIXES[oom_noise_desired]
+        prefix_center_got = offsetGot[-1-len(UNIT)]
+        prefix_center_desired = formatter.ENG_PREFIXES[oom_center_desired]
+        assert prefix_noise_desired == prefix_noise_got
+        assert prefix_center_desired == prefix_center_got
```
I'd just inline these; they aren't used elsewhere:
```suggestion
        assert formatter.ENG_PREFIXES[oom_noise_desired] == offsetGot[2]
        assert formatter.ENG_PREFIXES[oom_center_desired] == offsetGot[-1-len(UNIT)]
```

doronbehar replied:
> I'd just inline these; they aren't used elsewhere:

I thought about this too, but while debugging the tests with `pytest --showlocals`, I hope it is not a deal breaker to keep those, as I plan to maintain these tests and this feature in the future. I added a comment explaining that.

doronbehar replied:
Same here.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1594,6 +1594,79 @@ def test_engformatter_usetex_useMathText():
+        assert offsetGot == ""
+        # Make sure the ticks contain now the prefixes
+        for tick in ticksGot:
+            # 0 is zero on all orders of magnitudes, no
```
This sentence appears to be incomplete?

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1594,6 +1594,79 @@ def test_engformatter_usetex_useMathText():
+        # Make sure the ticks contain now the prefixes
+        for tick in ticksGot:
+            # 0 is zero on all orders of magnitudes, no
+            if tick[0] == "0":
+                prefixIdx = 0
+            else:
+                prefixIdx = oom_noise_desired
+            assert tick.endswith(formatter.ENG_PREFIXES[prefixIdx] + UNIT)
```
```suggestion
            index = 0 if tick[0] == "0" else oom_noise_desired
            assert tick.endswith(formatter.ENG_PREFIXES[index] + UNIT)
```

doronbehar replied:
Thanks for the suggestion, with your permission, again for `pytest --showlocals` I kept the `prefix_idx` variable name, I hope it is OK.

doronbehar replied:
And here.

QuLogic reviewed (commented):
Sorry, forgot to check the tests.

doronbehar commented on doc/users/next_whats_new/engformatter_offset.rst:
```diff
@@ -0,0 +1,14 @@
+:class:`matplotlib.ticker.EngFormatter` now computes offset by default
+----------------------------------------------------------------------
+
+:class:`matplotlib.ticker.EngFormatter` has gained the ability to show an
+offset text near the axis. With shared logic with
+:class:`matplotlib.ticker.ScalarFormatter`, it is capable of deciding whether
+the data qualifies having an offset and show it with an appropriate SI quantity
+prefix, and with the supplied ``unit``.
```
I'm still not fluent in Rst and since I'm also not sure what were your intentions I'll ask: You wanted to remove the links? Why? Also, if you want it to simply result in `<code>` formatted text, why not use double backticks?

QuLogic replied:
The suggested text is still a reference, and is less clutter for reading the raw text; code is double backticks, and unrelated.

doronbehar commented:
Thank you @QuLogic for the detailed review! I don't have time now to respond to all of your suggestions, so I started with the easier parts, documentation and comments.

doronbehar commented:
OK so last night I struggled a bit running the tests locally, so I hoped my changes would just work. Now I finally checked locally before pushing. I also responded to all other comments, and the topics left to discuss are the local variables helping to debug with `pytest --showlocals`, and meaning of `oom_` variables in the `test_engformatter_offset_oom`, and:

```rst
`matplotlib.ticker.EngFormatter`
```

v.s

```rst
:class:`matplotlib.ticker.EngFormatter`
```

In the docs (not in the titles of documents - that's understandable).

doronbehar commented:
I don't understand the circle CI errors... They don't seem related to my changes.

QuLogic commented:
> I don't understand the circle CI errors... They don't seem related to my changes. Same with the errors on Windows and MacOS on Azure.

It appears you have rebased against a very old commit instead of the current `main`.

doronbehar commented:
Rebased.

doronbehar commented:
Less errors now, but I see now:

```
Path does not exist: /Users/runner/work/1/s/result_images
```

ksunden commented on doc/users/next_whats_new/engformatter_offset.rst:
```diff
@@ -0,0 +1,13 @@
+``matplotlib.ticker.EngFormatter`` now computes offset by default
+-----------------------------------------------------------------
```
```suggestion
``matplotlib.ticker.EngFormatter`` now can compute offsets 
----------------------------------------------------------
```

(This behavior is off by default)

doronbehar commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -1591,6 +1591,79 @@ def test_engformatter_usetex_useMathText():
+    # only makes sure that offset text and the ticks gets the correct unit
+    # prefixes and the ticks.
+    if formatter.offset:
+        # These prefix_ variables are used only once, so we could have inlined
+        # them all, but it is more comfortable in case of tests breakages to
+        # view their values with pytest --showlocals.
```
> That pytest's assertion rewriting isn't showing you a breakdown of the results is a bug, and `--showlocals` is a reasonable workaround, but I wouldn't leave a comment about it.

What do you mean by pytest's assertion rewriting?

doronbehar commented:
The mypy CI error doesn't seem related to the PR...

ksunden commented:
Mypy passes locally, CI failure was fixed by #29014, would need a rebase to resolve on CI, butr since it passes locally and the failure is understood, I'm satisfied.

doronbehar commented:
Thank you for your corporation :).
